sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotCheckZeroSizeCollectionBase.cs

```diff
@@ -0,0 +1,146 @@
+        protected abstract TExpressionSyntax GetRightNode(TBinaryExpressionSyntax binaryExpression);
+        protected abstract ISymbol GetSymbol(SyntaxNodeAnalysisContext context, TExpressionSyntax expression);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer,
+                c =>
+                {
+                    var binaryExpression = (TBinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, GetLeftNode(binaryExpression), GetRightNode(binaryExpression));
+                }, GreaterThanOrEqualExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer,
+                c =>
+                {
+                    var binaryExpression = (TBinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, GetRightNode(binaryExpression), GetLeftNode(binaryExpression));
+                }, LessThanOrEqualExpression);
+        }
+
+        private void CheckCondition(SyntaxNodeAnalysisContext context, TExpressionSyntax expressionValueNode,
+            TExpressionSyntax constantValueNode)
+        {
+            if (!IsConstantZero(context, constantValueNode))
+            {
+                return;
+            }
+
+            ISymbol symbol = GetSymbol(context, expressionValueNode);
+            if (symbol == null)
+            {
+                return;
+            }
+
+            string symbolType = GetDeclaringTypeName(symbol);
+            if (symbolType == null)
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(Diagnostic.Create(Rule, context.Node.GetLocation(),
+                symbol.Name.ToLowerInvariant(), symbolType));
+        }
+
+        private static string GetDeclaringTypeName(ISymbol symbol)
+        {
+            if (IsArrayLengthProperty(symbol))
+            {
+                return "Array";
+            }
+
+            if (IsEnumerableCountMethod(symbol))
+            {
+                return "IEnumerable<T>";
+            }
+
+            if (IsCollectionProperty(symbol))
+            {
+                return "ICollection";
+            }
+
+            return null;
+        }
+
+        protected abstract TExpressionSyntax RemoveParentheses(TExpressionSyntax expression);
+
+        private bool IsConstantZero(SyntaxNodeAnalysisContext context, TExpressionSyntax expression)
+        {
+            var constantExpressionNode = RemoveParentheses(expression);
+            var constant = context.SemanticModel.GetConstantValue(constantExpressionNode);
+            if (!constant.HasValue)
+            {
+                return false;
+            }
+
+            return (constant.Value is int) && (int)constant.Value == 0;
+        }
+
+        private static bool IsEnumerableCountMethod(ISymbol symbol)
```

Comment:
Don't we already have those three methods in KnownMethods?

Reply:
We have IsEnumerableCountMethod, this one just adds a cast.
For the other two (IsArrayLengthProperty and IsCollectionProperty), we decided not to add them, as these are properties..